src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -23332,6 +23334,339 @@ struct DynamicPadToPad
+        return Bounds(-one, one);
+      }
+
+      auto optional_bounds = getBounds(boundsMap, op->getOperand(0));
+      if (!optional_bounds.has_value())
+        return std::nullopt;
+      auto bounds = optional_bounds.value();
+
+      if (auto negOp = dyn_cast<stablehlo::NegOp>(op)) {
+        return Bounds(-bounds.max, -bounds.min);
+      } else if (auto absOp = dyn_cast<stablehlo::AbsOp>(op)) {
+        APInt zero(bitWidth, 0, true);
+        if (bounds.min.sge(zero)) { // all positive
+          return Bounds(bounds.min, bounds.max);
+        } else if (bounds.max.sle(zero)) { // all negative
+          return Bounds(-bounds.max, -bounds.min);
+        } else {
+          APInt newMax = bounds.max.abs().sgt(-bounds.min.abs())
+                             ? bounds.max.abs()
+                             : (-bounds.min).abs();
+          return Bounds(zero, newMax);
+        }
+      }
+
+      return std::nullopt;
+    }
+
+    if (op->getNumOperands() == 2) {
+      auto optional_lhs_bounds = getBounds(boundsMap, op->getOperand(0));
+      auto optional_rhs_bounds = getBounds(boundsMap, op->getOperand(1));
+
+      if (!optional_lhs_bounds.has_value() || !optional_rhs_bounds.has_value())
+        return std::nullopt;
+
+      auto lhs_bounds = optional_lhs_bounds.value();
+      auto rhs_bounds = optional_rhs_bounds.value();
+
+      if (auto addOp = dyn_cast<stablehlo::AddOp>(op)) {
+        return Bounds(lhs_bounds.min + rhs_bounds.min,
+                      lhs_bounds.max + rhs_bounds.max);
+      } else if (auto subOp = dyn_cast<stablehlo::SubtractOp>(op)) {
+        return Bounds(lhs_bounds.min - rhs_bounds.max,
+                      lhs_bounds.max - rhs_bounds.min);
+      } else if (auto mulOp = dyn_cast<stablehlo::MulOp>(op)) {
+        auto p1 = lhs_bounds.min * rhs_bounds.min;
+        auto p2 = lhs_bounds.min * rhs_bounds.max;
+        auto p3 = lhs_bounds.max * rhs_bounds.min;
+        auto p4 = lhs_bounds.max * rhs_bounds.max;
+        return Bounds(min(min(p1, p2), min(p3, p4)),
+                      max(max(p1, p2), max(p3, p4)));
+      }
+    }
+
+    // TODO: other common ops like divide, remainder, clamp, etc.
+
+    return std::nullopt;
+  }
+
+  bool rewriteRemOp(PatternRewriter &rewriter, stablehlo::RemOp remOp,
+                    const DenseMap<Value, Bounds> &boundsMap) const {
+    auto optional_lhs_bounds = getBounds(boundsMap, remOp.getLhs());
+    auto optional_rhs_bounds = getBounds(boundsMap, remOp.getRhs());
+
+    if (!optional_lhs_bounds.has_value() || !optional_rhs_bounds.has_value())
+      return false;
+
+    auto lhs_bounds = optional_lhs_bounds.value();
+    auto rhs_bounds = optional_rhs_bounds.value();
+
+    if (lhs_bounds.min.sle(rhs_bounds.min) &&
+        rhs_bounds.max.sle(lhs_bounds.max)) {
+      auto lhs = remOp.getLhs();
+      rewriter.replaceOp(remOp, lhs);
+      return true;
+    }
+
+    return false;
+  };
+
+  bool rewriteCompareOp(PatternRewriter &rewriter,
```

Comment:
this seems somewhat (but not wholly) similar to: https://github.com/EnzymeAD/Enzyme-JAX/blob/9861ed866a3c9a64c860a50583615d3846ea337f/src/enzyme_ad/jax/Passes/AffineCFG.cpp#L5941